### ISSUE TITLE
Associations DSLs don't respect attribute order

### ISSUE THREAD

**Issue by lukasfroehlich1** [ElMassimo/oj_serializers#4]
The association DSL correctly serializes, but doesn't follow the same ordering conventions of the other attributes DSLs (attributes, serializer_attributes, etc). This results in all association DSLs always appearing at the end of the serialized JSON.

# Example

## Serializer

```
# frozen_string_literal: true

class AlbumSerializer < BaseSerializer
  attributes :id

  has_many :songs, serializer: SongSerializer

  attributes :genres
end
```

## Expected

```
{
  id: 123,
  songs: {...},
  genres: [...]
}
```

## Actual

```
{
  id: 123,
  genres: [...],
  songs: {...}
}
```

Happy to take a stab at fixing this if you point me in the right direction. Great library :). Thanks!

**Comment by ElMassimo** [comment]
Hi Lukas!

This is intentional, to simplify the implementation, as associations are stored in a separate data structure in the serializer schema.

Typically you wouldn't care about the order of a JSON structure, can you share more about your use case?

**Comment by lukasfroehlich1** [comment]
Thanks for the response. Since we're migrating over to OJ serializer we'd like to maintain the serialization order from before. While technically the order shouldn't matter, we'd like it to remain the same for our customers.

Happy to make a fork that does support association serialization order. Association DSL is not only faster but it makes the serializer more concise. Could you point me in the direction of what would need to change?

**Comment by ElMassimo** [comment]
You would need to combine [the `_attributes` and `_associations`](https://github.com/ElMassimo/oj_serializers/blob/master/lib/oj_serializers/serializer.rb#L357-L363) structures, so that the order is preserved, changing [these two loops](https://github.com/ElMassimo/oj_serializers/blob/master/lib/oj_serializers/serializer.rb#L375-L386) with a single one.

To do that, you will probably need to store a `type` value in the options when adding attributes or associations, so that you can later differentiate it when iterating that structure.

In the future I might revisit the internals of the implementation, as serializing in the same order as attributes are defined makes sense, but for now I'd suggest forking.

Feel free to open a _Discussion_ if you have any feedback or additional questions!

**Comment by ElMassimo** [comment]
@lukasfroehlich1 This is now solved since `oj_serializers-2.0.0`.
